### PR TITLE
Add validation for single statement lexical declarations

### LINKED ISSUE #3275: Assertion ECMA_PROPERTY_VALUE_PTR (property_p)->value == ECMA_VALUE_UNINITIALIZED in vm_loop
###### JerryScript revision
f93fa98a

###### Build platform
Linux-4.15.0-62-generic-x86_64-with-Ubuntu-18.04-bionic

###### Build steps
```
./tools/build.py --clean --debug --compile-flag=-fsanitize=address \
--compile-flag=-m32 --compile-flag=-fno-omit-frame-pointer \
--compile-flag=-fno-common --compile-flag=-g \
--strip=off --system-allocator=on --logging=on \
--linker-flag=-fuse-ld=gold --error-messages=on --profile=es2015-subset
```

###### Test case
```javascript
for ( i in this ) class ${ } eval ( )
```

###### Output
```text
ICE: Assertion 'ECMA_PROPERTY_VALUE_PTR (property_p)->value == ECMA_VALUE_UNINITIALIZED' failed at jerryscript/jerry-core/vm/vm.c(vm_loop):1331.
Error: ERR_FAILED_INTERNAL_ASSERTION
```

###### Backtrace
```text
bt
#0  0xf7fd5079 in __kernel_vsyscall ()
#1  0xf77fc832 in raise () from /lib/i386-linux-gnu/libc.so.6
#2  0xf77fdcc1 in abort () from /lib/i386-linux-gnu/libc.so.6
#3  0x5657ad5f in jerry_port_fatal (code=ERR_FAILED_INTERNAL_ASSERTION) at jerryscript/jerry-port/default/default-fatal.c:30
#4  0x565fdf83 in jerry_fatal (code=ERR_FAILED_INTERNAL_ASSERTION) at jerryscript/jerry-core/jrt/jrt-fatals.c:58
#5  0x565fdfc4 in jerry_assert_fail (assertion=0x566bcfa0 "ECMA_PROPERTY_VALUE_PTR (property_p)->value == ECMA_VALUE_UNINITIALIZED", file=0x566bc640 "jerryscript/jerry-core/vm/vm.c", function=0x5669d220 <__func__.6213.lto_priv.440> "vm_loop", line=1331) at jerryscript/jerry-core/jrt/jrt-fatals.c:82
#6  0x565de73f in vm_loop (frame_ctx_p=0xffffc440) at jerryscript/jerry-core/vm/vm.c:1331
#7  0x565ec334 in vm_execute (frame_ctx_p=0xffffc440, arg_p=0xffffc684, arg_list_len=0) at jerryscript/jerry-core/vm/vm.c:3727
#8  0x565ecc97 in vm_run (bytecode_header_p=0xf51028c0, this_binding_value=4126149459, lex_env_p=0xf5d00530, parse_opts=0, arg_list_p=0xffffc684, arg_list_len=0) at jerryscript/jerry-core/vm/vm.c:3853
#9  0x5662a284 in ecma_op_function_call (func_obj_p=0xf5f008b0, this_arg_value=72, arguments_list_p=0xffffc684, arguments_list_len=0) at jerryscript/jerry-core/ecma/operations/ecma-function-object.c:812
#10 0x565d9c42 in opfunc_call (frame_ctx_p=0xffffc6f0) at jerryscript/jerry-core/vm/vm.c:582
#11 0x565ec394 in vm_execute (frame_ctx_p=0xffffc6f0, arg_p=0x0, arg_list_len=0) at jerryscript/jerry-core/vm/vm.c:3733
#12 0x565ecc97 in vm_run (bytecode_header_p=0xf5300e10, this_binding_value=4126149459, lex_env_p=0xf5d007b0, parse_opts=0, arg_list_p=0x0, arg_list_len=0) at jerryscript/jerry-core/vm/vm.c:3853
#13 0x565d8eca in vm_run_global (bytecode_p=0xf5300e10) at jerryscript/jerry-core/vm/vm.c:282
#14 0x5666e510 in jerry_run (func_val=4126148835) at jerryscript/jerry-core/api/jerry.c:595
#15 0x5666ac86 in main (argc=2, argv=0xffffcb24) at jerryscript/jerry-main/main-unix.c:739
```

<sup>Found by [Fuzzinator](http://fuzzinator.readthedocs.io/) with [grammarinator](https://github.com/renatahodovan/grammarinator). </sup>

### LINKED ISSUE #3276: Assertion ecma_get_lex_env_type (frame_ctx_p->lex_env_p) == ECMA_LEXICAL_ENVIRONMENT_DECLARATIVE in vm_loop
###### JerryScript revision
f93fa98a

###### Build platform
Linux-4.15.0-62-generic-x86_64-with-Ubuntu-18.04-bionic

###### Build steps
```
./tools/build.py --clean --debug --compile-flag=-fsanitize=address \
--compile-flag=-m32 --compile-flag=-fno-omit-frame-pointer \
--compile-flag=-fno-common --compile-flag=-g \
--strip=off --system-allocator=on --logging=on \
--linker-flag=-fuse-ld=gold --error-messages=on --profile=es2015-subset
```

###### Test case
```javascript
with ({ }) class ${ $( ) { $ } } 
```

###### Output
```text
ICE: Assertion 'ecma_get_lex_env_type (frame_ctx_p->lex_env_p) == ECMA_LEXICAL_ENVIRONMENT_DECLARATIVE' failed at jerryscript/jerry-core/vm/vm.c(vm_loop):1324.
Error: ERR_FAILED_INTERNAL_ASSERTION
```

###### Backtrace
```text
bt
#0  0xf7fd5079 in __kernel_vsyscall ()
#1  0xf77fc832 in raise () from /lib/i386-linux-gnu/libc.so.6
#2  0xf77fdcc1 in abort () from /lib/i386-linux-gnu/libc.so.6
#3  0x5657ad5f in jerry_port_fatal (code=ERR_FAILED_INTERNAL_ASSERTION) at jerryscript/jerry-port/default/default-fatal.c:30
#4  0x565fdf83 in jerry_fatal (code=ERR_FAILED_INTERNAL_ASSERTION) at jerryscript/jerry-core/jrt/jrt-fatals.c:58
#5  0x565fdfc4 in jerry_assert_fail (assertion=0x566bc8a0 "ecma_get_lex_env_type (frame_ctx_p->lex_env_p) == ECMA_LEXICAL_ENVIRONMENT_DECLARATIVE", file=0x566bc640 "jerryscript/jerry-core/vm/vm.c", function=0x5669d220 <__func__.6213.lto_priv.440> "vm_loop", line=1324) at jerryscript/jerry-core/jrt/jrt-fatals.c:82
#6  0x565de57a in vm_loop (frame_ctx_p=0xffffc6f0) at jerryscript/jerry-core/vm/vm.c:1324
#7  0x565ec334 in vm_execute (frame_ctx_p=0xffffc6f0, arg_p=0x0, arg_list_len=0) at jerryscript/jerry-core/vm/vm.c:3727
#8  0x565ecc97 in vm_run (bytecode_header_p=0xf4f01ad0, this_binding_value=4126149459, lex_env_p=0xf5d007b0, parse_opts=0, arg_list_p=0x0, arg_list_len=0) at jerryscript/jerry-core/vm/vm.c:3853
#9  0x565d8eca in vm_run_global (bytecode_p=0xf4f01ad0) at jerryscript/jerry-core/vm/vm.c:282
#10 0x5666e510 in jerry_run (func_val=4126148835) at jerryscript/jerry-core/api/jerry.c:595
#11 0x5666ac86 in main (argc=2, argv=0xffffcb24) at jerryscript/jerry-main/main-unix.c:739
```

<sup>Found by [Fuzzinator](http://fuzzinator.readthedocs.io/) with [grammarinator](https://github.com/renatahodovan/grammarinator). </sup>

### PR DESCRIPTION
**Pull request by rerobika**
This patch fixes #3275 and fixes #3276.

JerryScript-DCO-1.0-Signed-off-by: Robert Fancsik frobert@inf.u-szeged.hu

### CODE REVIEW

**Comment by zherczeg** [comment]
One more thing: this is true for function statements except `if/else` statements, you could also add that as well.

File: jerry-core/parser/js/js-parser-statm.c [hunk 1]
```diff
@@ -477,6 +478,25 @@ parser_pop_block_context (parser_context_t *context_p) /**< context */
   parser_stack_iterator_init (context_p, &context_p->last_statement);
 } /* parser_pop_block_context */
 
+/**
+ * Validate lexical context for a declaration.
+ */
+static void
+parser_validate_lexical_context (parser_context_t *context_p) /**< context */
```
Comment (zherczeg) [inline jerry-core/parser/js/js-parser-statm.c]: I would not handle ES functions here.

Comment (rerobika) [reply jerry-core/parser/js/js-parser-statm.c]: Sure! I've updated the PR.

File: jerry-core/parser/js/js-parser-statm.c [hunk 2]
```diff
@@ -613,6 +634,15 @@ parser_parse_function_statement (parser_context_t *context_p) /**< context */
 
   JERRY_ASSERT (context_p->token.type == LEXER_KEYW_FUNCTION);
 
+#if ENABLED (JERRY_ES2015)
+  uint8_t flags = parser_statement_flags[context_p->stack_top_uint8];
+  if ((flags & PARSER_STATM_SINGLE_STATM)
+      && !(flags & PARSER_STATM_ALLOW_FUNC_DECL))
```
Comment (zherczeg) [inline jerry-core/parser/js/js-parser-statm.c]: This two check can be reduced to one with a PARSER_STATM_DISALLOW_FUNC_DECL flag.

**Pull request merged**